### PR TITLE
Add unsafe_project method to Subset interface and use it to clean up linalg.

### PR DESCRIPTION
**Pull request by duvenaud**
Almost every usage of `project` looked `fromJust $ project`.  This PR lets us replace all of those with `unsafe_project`, hopefully lessening the performance hit compared to unsafe indexing.

### CODE REVIEW

File: lib/prelude.dx [hunk 1]
```diff
@@ -895,79 +895,59 @@ instance {a n} [Ord a] Ord (n=>a)
...
 
 instance {q:Type} {i:q} [Ix q] Subset (RangeTo q i) q
   inject' = \(UnsafeMkRangeTo j). unsafe_from_ordinal _ j
-  project' = \j.
-    j' = ordinal j
-    i' = ordinal i
-    if j' > i'
-      then Nothing
-      else Just $ UnsafeMkRangeTo j'
+  can_project' = \j. ordinal j <= ordinal j
```
Comment (axch) [inline lib/prelude.dx:939]: Should there be an `ordinal i` here somewhere?  `ordinal j <= ordinal j` with two `j`s doesn't look right.  And a unit test that would catch the difference between `i` and `j`?

File: lib/prelude.dx [hunk 2]
```diff
@@ -2387,6 +2367,26 @@ def throw {a} (_:Unit) : {Except} a =
...
+'### Misc instances that require `todo`
+
+instance {a b} Subset a (a|b)
+  inject' = \x. Left x
+  can_project' = \x. case x of
+    Left  x -> True
+    Right x -> False
+  unsafe_project' = \x. case x of
+    Left  x -> x
+    Right x -> todo   -- Impossible
```
Comment (axch) [inline lib/prelude.dx:2379]: This situation is far from impossible: someone might call `unsafe_project` erroneously.

**Review by axch — COMMENTED** [review]
Introducing `unsafe_project` cleans up the example code nicely!

However, I have mixed feelings about factoring the interface as `can_project` and `unsafe_project`, as opposed to `project` and `unsafe_project`.  The former is more elegant, in that it avoids duplicating code; but the latter avoids duplicating work in the common case.

What do I mean by that?  If we define the interface as `can_project` and `unsafe_project`, then we define `project` as a function, whose body is
```
if can_project x
  then Just $ unsafe_project x
  else Nothing
```
However, `can_project` often does some work (like calling `ordinal` in the `RangeTo` instances, or scrutinizing the index in the `Either` instance).  It then throws almost all of the result of this work away by returning only a Bool value, and the `unsafe_project` instance has to repeat it.

Now, in the case of the instances we have now, it probably doesn't matter very much; and in the case where we inline both methods (which is presumably always) good CSE and case-of-known-constructor optimization should be able to eliminate the overhead.  But I worry that the interface bakes in this inefficiency, and in sufficiently complex cases CSE won't be able to fix it.

What do people think?  Should we accept this risk of inefficiency in order to factor the code into orthogonal pieces?  Should we maybe have both `can_project` and `project` in the interface, and add default methods (do we have those?) to fill them in in terms of each other?  Other ideas?

P.S.  This kind of inefficiency always happens.  Boolean values appearing in data structures, or functions that return a boolean value, are a code smell everywhere except as arguments to something like `filter` that really just makes a binary decision and can't use any other information.

**Review by apaszke — COMMENTED** [review]
I agree with Alexey! the only difference between `can_project` and `project` is that the second calls `unsafe_project` after performing the checks, so the duplication seems minimal.

File: lib/prelude.dx (hunk 2 above)
Comment (duvenaud) [reply lib/prelude.dx:2379]: Good point, this should be an `error`.

File: lib/prelude.dx (hunk 1 above)
Comment (duvenaud) [reply lib/prelude.dx:939]: Yes, this is clearly a bug, and shows that I didn't put in enough tests.

**Comment by duvenaud** [comment]
Thanks for the feedback, guys.  I Actually started with `project` and `unsafe_project` as you suggested.  I stopped because one currently can't call another method within the same instance definition.  But I can always simply define a little function at the top level and have both `project` and `unsafe_project` call it.  I'll make this change.

And thanks @axch for the lesson in design, I agree that functions returning a `Bool` smell like wasted computation.

File: lib/prelude.dx (hunk 1 above)
Comment (duvenaud) [reply lib/prelude.dx]: I restored all the old `project` instances back to their old selves.

**Comment by duvenaud** [comment]
Okay, I made the suggested changes, I think this is ready for another review or merging now.

File: lib/prelude.dx [hunk 3]
```diff
@@ -915,38 +910,65 @@ def inject {a_sub} (a:Type) [Subset a_sub a] (x:a_sub) : a =
...
-  can_project' = \j. ordinal j > ordinal i
+  project' = \j.
+    j' = ordinal j
+    i' = ordinal i
+    if j' <= i'
+      then Nothing
+      else Just $ UnsafeMkRangeFromExc $ unsafe_nat_diff j' (i' + 1)
   unsafe_project' = \j.
-    UnsafeMkRangeFromExc $ unsafe_nat_diff (ordinal j) (ordinal i + 1)
+    UnsafeMkRangeFromExc $ unsafe_nat_diff (ordinal j) ((ordinal i) + 1)
```
Comment (axch) [inline lib/prelude.dx:942]: Nit: You shouldn't need these parens.

File: lib/linalg.dx [hunk 4]
```diff
@@ -16,6 +15,8 @@ def refl_more {n} [Ix n] (i:n) : (i..) = unsafe_project _ i
 def upper_tri_diag {n v} (u:UpperTriMat n v) : n=>v = view i. u.i.(refl_more i)
 def lower_tri_diag {n v} (l:LowerTriMat n v) : n=>v = view i. l.i.(refl_less i)
 
+def is_last {n} [Ix n] (x:n) : Bool = ((ordinal x) + 1) == size n
```
Comment (axch) [inline lib/linalg.dx:18]: Nit: This should work without parens at all: `ordinal x + 1 == size n`, though it's up to you which form you consider the most readable.

File: lib/prelude.dx [hunk 5]
```diff
@@ -2387,6 +2389,25 @@ def throw {a} (_:Unit) : {Except} a =
 def assert (b:Bool) : {Except} Unit =
   if not b then throw ()
 
+'### Misc instances that require `todo`
```
Comment (axch) [inline lib/prelude.dx:2392]: This comment is not still true, is it?  There's no `todo` here now, right?

**Review by axch — CHANGES_REQUESTED** [review]
This is great!  Just a couple tiny nits and it's good to go!

File: lib/prelude.dx (hunk 3 above)
Comment (duvenaud) [reply lib/prelude.dx]: Fixed.

File: lib/linalg.dx (hunk 4 above)
Comment (duvenaud) [reply lib/linalg.dx]: Actually it does require the outer parens, but I got rid of the inner ones, thanks.

File: lib/prelude.dx (hunk 5 above)
Comment (duvenaud) [reply lib/prelude.dx]: Fixed, thanks.

**Review by axch — APPROVED** [review]

**Pull request merged**